HW2/P2/parallel_vector.pyx

```diff
@@ -94,22 +122,42 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
                                np.int32_t[:] dest,
                                int repeat,
                                int N):
-   cdef:
-       int idx, r
-       int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
-       omp_lock_t *locks = get_N_locks(num_locks)
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+    cdef:
+        int idx, r
+        int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
+        omp_lock_t *locks = get_N_locks(num_locks)
+        omp_lock_t *min_lock
+        omp_lock_t *max_lock
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for every N adjacent elements of counts
+    # to parallelize data movement.  Be sure to avoid deadlock, as well as
+    # double-locking.
+    ##########
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4):
+
+                # only modification compared to fine-grained locking is to divide indices by N
+                # so that adjacent cells map to the proper lock
+                min_lock = &locks[src[idx] / N] if src[idx] < dest[idx] else &locks[dest[idx] / N]
```

Comment:
Nice to see the same things as above coming here. Also comments throughout made this easy to follow.